### DISCUSSION TITLE
How to tell if Serial is fully initialized?

### DISCUSSION THREAD

**Discussion by joba-1** [earlephilhower/arduino-pico#1256] [Q&A] [answered]
Hi,

I noticed that my early boot messages are never displayed on Serial/USB of my Pico-W.
Messages seem to appear ~1s after boot.
I guess this is because the Pico first has to negotiate USB stuff with the host at each boot.

Is there a way to find out if a message made it to the wire or do I need to delay important messages for 1s and hope for the best?

**Comment by earlephilhower** [comment, marked as answer]
`while (!Serial);` will get you 80% of the way, but the problem is the OS and the Serial Monitor may still take 1-2s before they reconnect and anything send before then will drop.  There's nothing on the Pico side that can be done about that.

Something like a `delay(5000)` or a `while (!Serial); delay(2000);` might get you 99% of the way there.

**Reply by joba-1** [reply]
Thank you for the insight.

Fair enough, I guess, for a blink sketch a delay of several seconds and a 99% chance is good enough :)

But during development with frequent reboots this would drive me crazy.

I’ll try my luck with Serial1 and an external always on usb to serial adapter like ftdi or the picoprobe. Will be interesting to convince external libraries of that change…
